agent/engine/dockerauth/ecr.go

```diff
@@ -17,52 +17,98 @@ import (
 	"fmt"
 	"strings"
+	"sync"
+	"time"
 
 	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/async"
 	"github.com/aws/amazon-ecs-agent/agent/ecr"
 	ecrapi "github.com/aws/amazon-ecs-agent/agent/ecr/model/ecr"
+	"github.com/aws/amazon-ecs-agent/agent/utils"
 	"github.com/aws/aws-sdk-go/aws"
 	log "github.com/cihub/seelog"
 	docker "github.com/fsouza/go-dockerclient"
 )
 
+type cacheKey struct {
+	region           string
+	rolearn          string
+	registryID       string
+	endpointOverride string
+}
+
 type ecrAuthProvider struct {
-	authData *api.ECRAuthData
-	client   ecr.ECRClient
+	tokenCache async.Cache
+	factory    ecr.ECRFactory
+	cacheLock  sync.RWMutex
 }
 
-const proxyEndpointScheme = "https://"
+const (
+	roundtripTimeout      = 5 * time.Second
+	MinimumJitterDuration = 30 * time.Minute
+	proxyEndpointScheme   = "https://"
+)
+
+func (key *cacheKey) String() string {
+	return fmt.Sprintf("%s-%s-%s-%s", key.rolearn, key.region, key.registryID, key.endpointOverride)
+}
 
 // NewECRAuthProvider returns a DockerAuthProvider that can handle retrieve
 // credentials for pulling from Amazon EC2 Container Registry
-func NewECRAuthProvider(authData *api.ECRAuthData, clientFactory ecr.ECRFactory) DockerAuthProvider {
-	if authData == nil {
-		return &ecrAuthProvider{}
-	}
-	log.Tracef("Getting client in %s with endpoint %s", authData.Region, authData.EndpointOverride)
+func NewECRAuthProvider(ecrFactory ecr.ECRFactory, cache async.Cache) DockerAuthProvider {
 	return &ecrAuthProvider{
-		authData: authData,
-		client:   clientFactory.GetClient(authData.Region, authData.EndpointOverride),
+		tokenCache: cache,
+		factory:    ecrFactory,
 	}
 }
 
 // GetAuthconfig retrieves the correct auth configuration for the given repository
-func (authProvider *ecrAuthProvider) GetAuthconfig(image string) (docker.AuthConfiguration, error) {
-	if authProvider.authData == nil {
-		return docker.AuthConfiguration{}, fmt.Errorf("ecrAuthProvider cannot be used without AuthData")
+func (authProvider *ecrAuthProvider) GetAuthconfig(image string,
+	authData *api.ECRAuthData) (docker.AuthConfiguration, error) {
+	if authData == nil {
+		return docker.AuthConfiguration{}, fmt.Errorf("Missing AuthConfiguration data")
+	}
+
+	// First try to get the token from cache, if the token does not exist,
+	// then call ecr api to get the new token
+	key := cacheKey{
+		region:           authData.Region,
+		endpointOverride: authData.EndpointOverride,
+		registryID:       authData.RegistryID,
+	}
+	if !utils.ZeroOrNil(authData.GetPullCredentials()) {
+		key.rolearn = authData.GetPullCredentials().RoleArn
 	}
+	authProvider.cacheLock.RLock()
+	defer authProvider.cacheLock.RUnlock()
+
+	token, ok := authProvider.tokenCache.Get(key.String())
+	if ok {
+		cachedToken := token.(*ecrapi.AuthorizationData)
+		if authProvider.ISTokenValid(cachedToken) {
+			return extractToken(cachedToken)
+		}
+		authProvider.tokenCache.Delete(key.String())
```

Comment:
please add comments here explaining why it's ok to explicitly remove this from the cache.